[Feature] Add materialized view definition

ABingHuang opened the pull request:
## What type of PR is this：
- [x] bug
- [ ] feature
- [ ] enhancement
- [ ] others

## Which issues of this PR fixes ：
<!--
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #4787 

## Problem Summary(Required) ：
<!-- (Please describe the changes you have made. In which scenarios will this bug be triggered and what measures have you taken to fix the bug?) -->
1. add MaterializedView meta class
2. add MaterializedViewManager to manage the relation between base tables and MaterializedView

wanpengfei-git commented:
#### [FE PR Coverage check]

:disappointed: **fail** : 193 / 252 (76.59%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/Table.java|2|6|33.33%|
|:large_blue_circle:|com/starrocks/catalog/Database.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|151|205|73.66%|
|:large_blue_circle:|com/starrocks/catalog/PartitionInfo.java|1|1|100.00%|
|:large_blue_circle:|com/starrocks/alter/MaterializedViewManager.java|38|38|100.00%|

Astralidea commented on fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedView.java:
```diff
@@ -0,0 +1,494 @@
+    // TODO: now it is original definition sql
+    // for show create mv, constructing refresh job(insert into select)
+    @SerializedName(value = "viewDefineSql")
+    private String viewDefineSql;
```
This defineSql need to be expose for generate insert overwrite sql

ABingHuang replied:
ok. I will add one api

ABingHuang commented on be/test/storage/tablet_updates_test.cpp:
```diff
@@ -13,18 +13,18 @@
 #include "storage/empty_iterator.h"
-#include "storage/schema_change.h"
-#include "storage/union_iterator.h"
 #include "storage/kv_store.h"
```
it is added by clang format tool. I remove it in this pr.

ABingHuang commented on fe/fe-core/src/main/java/com/starrocks/alter/MaterializedViewManager.java:
```diff
@@ -0,0 +1,100 @@
+public class MaterializedViewManager {
+    private static MaterializedViewManager instance = new MaterializedViewManager();
+    // mv name -> materialized view
+    private Map<String, MaterializedView> materializedViews;
```
it can be get from catalog because materialized view is a type of table

HangyuanLiu commented on fe/fe-core/src/main/java/com/starrocks/catalog/OlapTable.java:
```diff
@@ -98,31 +100,49 @@ public enum OlapTableState {
+    @SerializedName(value = "bfFpp")
     protected double bfFpp;
 
+    @SerializedName(value = "colocateGroup")
```
I recommend adding only the properties we need for now

ABingHuang replied:
OK, I will remove the codes

ABingHuang commented on fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedView.java:
```diff
@@ -0,0 +1,214 @@
+
+    @Override
+    public void write(DataOutput out) throws IOException {
+        // write other fields in gson
```
it is an old comment, I will remove it.

ABingHuang committed:
Add materialized view definition

1. make MaterializedView extends OlapTable
2. add MaterializedViewManager to Catalog

ABingHuang committed:
modify OlapTable serialization by gson

wyb commented on fe/fe-core/src/main/java/com/starrocks/catalog/RangePartitionInfo.java:
```diff
@@ -48,10 +49,13 @@
     private List<Column> partitionColumns = Lists.newArrayList();
     // formal partition id -> partition range
+    @SerializedName(value = "idToRange")
     private Map<Long, Range<PartitionKey>> idToRange = Maps.newHashMap();
```
Currently, `Range` and `PartitionKey` json serialization is not supported.

wyb commented on fe/fe-core/src/main/java/com/starrocks/catalog/Table.java:
```diff
@@ -64,11 +65,16 @@ public enum TableType {
         HUDI,
         ODBC,
         JDBC,
+        MATERIALIZEDVIEW,
```
```suggestion
        MATERIALIZED_VIEW,
```

ABingHuang committed:
modify comment

ABingHuang committed:
fix Range<PartitionKey> serialization bug

ABingHuang committed:
fix compile

ABingHuang commented:
run starrocks_fe_unittest

wanpengfei-git commented:
#### [FE PR Coverage check]

:disappointed: **fail** : 183 / 235 (77.87%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/Table.java|2|6|33.33%|
|:large_blue_circle:|com/starrocks/catalog/Database.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|50|75|66.67%|
|:large_blue_circle:|com/starrocks/catalog/RangePartitionInfo.java|70|91|76.92%|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|15|15|100.00%|
|:large_blue_circle:|com/starrocks/persist/gson/GsonUtils.java|5|5|100.00%|
|:large_blue_circle:|com/starrocks/catalog/PartitionInfo.java|3|3|100.00%|
|:large_blue_circle:|com/starrocks/alter/MaterializedViewManager.java|36|36|100.00%|

ABingHuang committed:
optimize serialization

wanpengfei-git commented:
#### [FE PR Coverage check]

:heart_eyes: **pass** : 146 / 177 (82.49%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/Table.java|2|6|33.33%|
|:large_blue_circle:|com/starrocks/catalog/Database.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/server/GlobalStateMgr.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|50|75|66.67%|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|15|15|100.00%|
|:large_blue_circle:|com/starrocks/catalog/RangePartitionInfo.java|33|33|100.00%|
|:large_blue_circle:|com/starrocks/persist/gson/GsonUtils.java|5|5|100.00%|
|:large_blue_circle:|com/starrocks/catalog/PartitionInfo.java|3|3|100.00%|
|:large_blue_circle:|com/starrocks/alter/MaterializedViewManager.java|36|36|100.00%|

ABingHuang committed:
remove null check

kangkaisen commented:
I'm reviewing this PR, please don't merge it.

kangkaisen commented on fe/fe-core/src/main/java/com/starrocks/alter/MaterializedViewManager.java:
```diff
@@ -0,0 +1,94 @@
+
+    // table id -> mv set
+    // every table that has mvs will be recorded here
+    private Map<Long, Set<MaterializedView>> materializedViewsForTable;
```
how about directly push the MV ids to OlapTable?   then we needn't this map.

ABingHuang replied:
I pretend to add the relationship here. I do not want to add fields to OlapTable for back-compatible reasons.

ABingHuang replied:
I will move the relations to OlapTable and delete MaterializedViewManager

kangkaisen commented on fe/fe-core/src/main/java/com/starrocks/alter/MaterializedViewManager.java:
```diff
@@ -0,0 +1,94 @@
+// 1. manage the relations between manterialized view and base tables
+public class MaterializedViewManager {
+    // mv name -> materialized view
+    private Map<String, MaterializedView> materializedViews;
```
why do need this key is MV name? when we rewrite the query, how do we use it?

ABingHuang replied:
It is not very useful. I will remove it.
If needed to get mv from name, it can be done by Catalog

kangkaisen commented on fe/fe-core/src/main/java/com/starrocks/catalog/MaterializedView.java:
```diff
@@ -0,0 +1,214 @@
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
override this method is no meaning.

ABingHuang replied:
ok, I will remove it.

kangkaisen commented on fe/fe-core/src/main/java/com/starrocks/catalog/RangePartitionInfo.java:
```diff
@@ -48,12 +54,25 @@
     private Map<Long, Range<PartitionKey>> idToTempRange = Maps.newHashMap();
 
+    // partitionId -> serialized Range<PartitionKey>
+    // because Range<PartitionKey> and PartitionKey can not be serialized by gson
```
```
because Range<PartitionKey> and PartitionKey can not be serialized by gson
```
why? I think you could change it.

ABingHuang replied:
Range is defined in guava. and I can not modify it. And more, If I add another data structure to replace Range, all partition related logic should be modified. So I use the method in this pr, solve this problem when serialization and deserialization.

kangkaisen commented on gensrc/thrift/Descriptors.thrift:
```diff
@@ -240,6 +240,10 @@ struct TOlapTable {
     1: required string tableName
 }
 
+struct TMaterializedView {
```
why do need pass the mv name info to BE?  when do we need this thrift info in BE?

ABingHuang replied:
I will remove this structure.

ABingHuang committed:
fix comment

ABingHuang committed:
fix comment

ABingHuang committed:
remove ut

ABingHuang committed:
fix

ABingHuang committed:
fix toThrift

ABingHuang committed:
fix compile

ABingHuang committed:
fix ut

wanpengfei-git commented:
#### [FE PR Coverage check]

:disappointed: **fail** : 104 / 142 (73.24%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/Table.java|2|6|33.33%|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|46|75|61.33%|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|17|22|77.27%|
|:large_blue_circle:|com/starrocks/catalog/RangePartitionInfo.java|31|31|100.00%|
|:large_blue_circle:|com/starrocks/persist/gson/GsonUtils.java|5|5|100.00%|
|:large_blue_circle:|com/starrocks/catalog/PartitionInfo.java|3|3|100.00%|

kangkaisen commented:
run starrocks_be_unittest

ABingHuang committed:
add ut

ABingHuang committed:
fix ut

wanpengfei-git commented:
#### [FE PR Coverage check]

:heart_eyes: **pass** : 122 / 145 (84.14%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/catalog/MaterializedView.java|55|78|70.51%|
|:large_blue_circle:|com/starrocks/catalog/Table.java|6|6|100.00%|
|:large_blue_circle:|com/starrocks/catalog/OlapTable.java|22|22|100.00%|
|:large_blue_circle:|com/starrocks/catalog/RangePartitionInfo.java|31|31|100.00%|
|:large_blue_circle:|com/starrocks/persist/gson/GsonUtils.java|5|5|100.00%|
|:large_blue_circle:|com/starrocks/catalog/PartitionInfo.java|3|3|100.00%|
